File: app/src/main/java/rocks/poopjournal/todont/Fragments/YearlyFragment.java
package rocks.poopjournal.todont.Fragments;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import androidx.fragment.app.Fragment;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.components.Description;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.data.PieEntry;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

import rocks.poopjournal.todont.Db_Controller;
import rocks.poopjournal.todont.Helper;
import rocks.poopjournal.todont.R;

import static java.util.Calendar.getInstance;

public class YearlyFragment extends Fragment {
    Calendar c = getInstance();
    Calendar d = getInstance();
    Calendar e = getInstance();
    SimpleDateFormat df;
    Db_Controller db;
    TextView year, mostavoided, leastavoided, daterange;
    Button btnbefore, btnafter;
    String formattedDate;
    String getmostAvoidedHabit;
    SharedPreferences prefs;
    String checkDate;
    PieChart pieChart;
    String curmonth;
    String[] splitteddate;
    String b_month;
    int b_year;
    double habitsSize, avoidedSize;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_yearly, container, false);
        db = new Db_Controller(getActivity(), "", null, 2);
        df = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
        prefs = getContext().getSharedPreferences("MyPrefs", Context.MODE_PRIVATE);
        pieChart = view.findViewById(R.id.pieChart);
        checkDate = prefs.getString("InitialDate", "");
        Helper.SelectedButtonOfLogTab = 3;
        year = view.findViewById(R.id.year);
        pieChart = view.findViewById(R.id.pieChart);
        btnbefore = view.findViewById(R.id.before);
        btnafter = view.findViewById(R.id.after);
        btnbefore.setBackgroundResource(R.drawable.ic_backarrow);
        btnafter.setBackgroundResource(R.drawable.ic_nextarrow);
        daterange = view.findViewById(R.id.daterange);
        mostavoided = view.findViewById(R.id.mostavoided);
        leastavoided = view.findViewById(R.id.leastavoided);
        db.show_habits_data();
        final String cur_date = df.format(c.getTime());
        splitteddate = cur_date.split("-");
        b_year = Integer.parseInt(splitteddate[0]);
        year.setText("" + splitteddate[0]);
        String newDate = splitteddate[0] + "-01-" + "-01";
        String newDate1 = splitteddate[0] + "-12-" + "-31";
        Log.d("splitteddate333 ", "" + newDate + " * " + newDate1);
        String avoided = db.getYearlyAvoidedData(newDate, newDate1);
        String done = db.getYearlyDoneData(newDate, newDate1);
        mostavoided.setText("" + avoided);
        leastavoided.setText("" + done);
        habitsSize = Helper.habitsdata.size() * 365;
        avoidedSize = Helper.avoidedyearlydata.size();
        double per = (avoidedSize / habitsSize) * 100;
        Log.d("percentage", "per : " + per + " yearly" + Helper.avoidedyearlydata.size() + "total" + Helper.habitsdata.size());
        funcPieChart((int) per);
        String[] checkdatea = checkDate.split("-");
        if (year.getText().toString().equals(checkdatea[0])) {
            btnbefore.setEnabled(false);
        }
        else{
            btnbefore.setEnabled(true);
            btnafter.setEnabled(true);
        }
        if(year.getText().toString().equals(splitteddate[0])){
            btnafter.setEnabled(false);
        }
        else{
            btnafter.setEnabled(true);
        }

        btnbefore.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                btnafter.setEnabled(true);
                b_year = b_year - 1;

                year.setText("" + b_year);
                btnbefore.setBackgroundResource(R.drawable.ic_backarrowpressed);
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        btnbefore.setBackgroundResource(R.drawable.ic_backarrow);
                    }
                }, 100);
                String[] checkdatearray = checkDate.split("-");
                b_year = Integer.parseInt(year.getText().toString());
                Log.d("fff", "" + year.getText().toString() + " ggg  " + checkdatearray[0]);
                if ((year.getText().toString()).equals(checkdatearray[0])) {
                    btnbefore.setEnabled(false);
                }
                String newDate = b_year + "-01" + "-01-";
                String newDate1 = b_year + "-12" + "-31-";
                Log.d("sssssssss", "" + newDate + " * " + newDate1);
                String avoided = db.getYearlyAvoidedData(newDate, newDate1);
                String done = db.getYearlyDoneData(newDate, newDate1);
                mostavoided.setText("" + avoided);
                leastavoided.setText("" + done);
                habitsSize = Helper.habitsdata.size() * 365;
                avoidedSize = Helper.avoidedyearlydata.size();
                double per = (avoidedSize / habitsSize) * 100;
                Log.d("percentage", "per : " + per + " monthly" + Helper.avoidedyearlydata.size() + "total" + Helper.habitsdata.size());
                funcPieChart((int) per);
            }
        });
        btnafter.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                btnbefore.setEnabled(true);
                b_year = b_year + 1;
                year.setText("" + b_year);
                btnbefore.setBackgroundResource(R.drawable.ic_backarrowpressed);
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        btnbefore.setBackgroundResource(R.drawable.ic_backarrow);
                    }
                }, 100);
                String[] checkdatearray = checkDate.split("-");
                b_year = Integer.parseInt(year.getText().toString());
                final String cur_date = df.format(c.getTime());
                splitteddate = cur_date.split("-");
                if ((year.getText().toString()).equals(splitteddate[0])) {
                    btnafter.setEnabled(false);
                }
                String newDate = b_year + "-01" + "-01-";
                String newDate1 = b_year + "-12" + "-31-";
                String avoided = db.getYearlyAvoidedData(newDate, newDate1);
                String done = db.getYearlyDoneData(newDate, newDate1);
                mostavoided.setText("" + avoided);
                leastavoided.setText("" + done);
                habitsSize = Helper.habitsdata.size() * 365;
                avoidedSize = Helper.avoidedyearlydata.size();
                double per = (avoidedSize / habitsSize) * 100;
                Log.d("percentage", "per : " + per + " monthly" + Helper.avoidedyearlydata.size() + "total" + Helper.habitsdata.size());
                funcPieChart((int) per);
            }
        });

        return view;
    }


    public void funcPieChart(int avoidedPer) {
        pieChart.setUsePercentValues(true);
        List<PieEntry> value = new ArrayList<>();
        value.add(new PieEntry((float) avoidedPer, "Avoided"));
        value.add(new PieEntry((float) (100.0 - avoidedPer), "Habits"));
        PieDataSet pieDataSet = new PieDataSet(value, "");
        pieDataSet.setValueTextColor(Color.WHITE);
        PieData pieData = new PieData(pieDataSet);
        pieChart.setData(pieData);
        pieDataSet.setColors(new int[]{Color.parseColor("#FFAF01"), Color.parseColor("#26272c")});
        Legend legend = pieChart.getLegend();
        legend.setEnabled(false);
        Description description = new Description();
        description.setText("");
        pieChart.setDescription(description);
        pieChart.setHoleRadius(50f);
        pieChart.setTransparentCircleRadius(50f);
        pieChart.setHoleColor(getResources().getColor(R.color.backgroundcolor));
        pieChart.animateXY(1000, 1000);
    }
}

File: app/src/main/java/rocks/poopjournal/todont/fragments/YearlyFragment.kt
package rocks.poopjournal.todont.fragments

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.components.Description
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import rocks.poopjournal.todont.Helper
import rocks.poopjournal.todont.R
import rocks.poopjournal.todont.utils.Constants
import rocks.poopjournal.todont.utils.DatabaseUtils
import rocks.poopjournal.todont.utils.SharedPrefUtils
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

class YearlyFragment : Fragment() {

    private lateinit var calendar: Calendar
    private lateinit var dbHelper: DatabaseUtils
    private lateinit var dateFormat: SimpleDateFormat
    private lateinit var preferences: SharedPrefUtils

    private var pieChart: PieChart? = null
    private var currentYear: Int = 0
    private var initialDate: String? = null

    private lateinit var yearText: TextView
    private lateinit var mostAvoidedText: TextView
    private lateinit var leastAvoidedText: TextView
    private lateinit var dateRangeText: TextView
    private lateinit var btnPrevious: ImageView
    private lateinit var btnNext: ImageView

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.fragment_yearly, container, false)

        initializeComponents(view)
        setupInitialData()
        updateYearData()

        btnPrevious.setOnClickListener {
            updateYear(-1)
        }
        btnNext.setOnClickListener {
            updateYear(1)
        }

        return view
    }

    private fun initializeComponents(view: View) {
        dbHelper = DatabaseUtils(requireContext())
        dateFormat = SimpleDateFormat(Constants.DATE_FORMAT, Locale.getDefault())
        preferences = SharedPrefUtils(requireContext())

        yearText = view.findViewById(R.id.year)
        mostAvoidedText = view.findViewById(R.id.mostavoided)
        leastAvoidedText = view.findViewById(R.id.leastavoided)
        dateRangeText = view.findViewById(R.id.daterange)
        btnPrevious = view.findViewById(R.id.before)
        btnNext = view.findViewById(R.id.after)
        pieChart = view.findViewById(R.id.pieChart)

        btnPrevious.setBackgroundResource(R.drawable.ic_backarrow)
        btnNext.setBackgroundResource(R.drawable.ic_nextarrow)
    }

    private fun setupInitialData() {
        calendar = Calendar.getInstance()
        currentYear = calendar.get(Calendar.YEAR)
        yearText.text = currentYear.toString()

        initialDate = preferences.getString("InitialDate", getTodayDate())
        Helper.SelectedButtonOfLogTab = 3
    }

    private fun updateYear(offset: Int) {
        currentYear += offset
        yearText.text = currentYear.toString()

        Handler().postDelayed({
            btnPrevious.setBackgroundResource(R.drawable.ic_backarrow)
        }, 100)

        updateYearData()

//        btnPrevious.isEnabled = initialDate?.split("-")?.get(0)?.toInt() != currentYear
//        btnNext.isEnabled = currentYear != Calendar.getInstance().get(Calendar.YEAR)
    }

    private fun updateYearData() {
        val startDate = "$currentYear-01-01"
        val endDate = "$currentYear-12-31"

        val avoided = dbHelper.getYearlyAvoidedData(startDate, endDate)
        val done = dbHelper.getYearlyDoneData(startDate, endDate)

        if(avoided.isNotEmpty()){
            mostAvoidedText.text = dbHelper?.getHabitById(avoided.toInt())?.name
            mostAvoidedText.visibility=View.VISIBLE
        }else{
            mostAvoidedText.visibility=View.GONE
        }

        if(done.isNotEmpty()){
            leastAvoidedText.text = dbHelper?.getHabitById(done.toInt())?.name
            leastAvoidedText.visibility=View.VISIBLE
        }else{
            leastAvoidedText.visibility=View.GONE
        }


        val totalHabits = dbHelper.getHabitsCount() * 365.0
        val avoidedCount = dbHelper.getYearlyAvoidedDataList(startDate, endDate).size.toDouble()
        val percentage = (avoidedCount / totalHabits) * 100

        setupPieChart(percentage.toInt())
    }

    private fun setupPieChart(avoidedPercentage: Int) {
        pieChart?.apply {
            setUsePercentValues(true)

            val entries = listOf(
                PieEntry(avoidedPercentage.toFloat(), getString(R.string.avoided)),
                PieEntry((100 - avoidedPercentage).toFloat(), getString(R.string.habits))
            )
            val dataSet = PieDataSet(entries, "").apply {
                valueTextColor = Color.WHITE
                val lightSurface=TypedValue()
                requireContext().theme.resolveAttribute(R.attr.colorOnBackground3,lightSurface,true)
                val primaryColor=TypedValue()
                requireContext().theme.resolveAttribute(R.attr.colorAccent,primaryColor,true)
                setColors(primaryColor.data, lightSurface.data)
            }

            data = PieData(dataSet)
            description = Description().apply { text = "" }
            holeRadius = 50f
            transparentCircleRadius = 50f
            val typedValue = TypedValue()
            requireContext().theme.resolveAttribute(R.attr.colorBackground, typedValue, true)
            setHoleColor(typedValue.data)
            animateXY(1000, 1000)
        }
    }
    private fun getTodayDate(): String {
        var date: Date = Calendar.getInstance().time
        val dateFormat: SimpleDateFormat = SimpleDateFormat(Constants.DATE_FORMAT, Locale.getDefault())
        var formattedDate: String = dateFormat.format(date)
        //Calendar.getInstance().timeInMillis.toString()
        return formattedDate
    }
}
